include/llvm/IR/ValueHandle.h

```diff
@@ -275,46 +274,65 @@ struct isPodLike<AssertingVH<T> > {
 #endif
 };
 
-
 /// \brief Value handle that tracks a Value across RAUW.
 ///
 /// TrackingVH is designed for situations where a client needs to hold a handle
 /// to a Value (or subclass) across some operations which may move that value,
 /// but should never destroy it or replace it with some unacceptable type.
 ///
-/// It is an error to do anything with a TrackingVH whose value has been
-/// destroyed, except to destruct it.
-///
 /// It is an error to attempt to replace a value with one of a type which is
 /// incompatible with any of its outstanding TrackingVHs.
-template<typename ValueTy>
-class TrackingVH : public ValueHandleBase {
-  void CheckValidity() const {
-    Value *VP = ValueHandleBase::getValPtr();
-
-    // Null is always ok.
-    if (!VP) return;
+///
+/// It is an error to read from a TrackingVH that does not point to a valid
+/// value.  A TrackingVH is said to not point to a valid value if either it
+/// hasn't yet been assigned a value yet or because the value it was tracking
+/// has since been deleted.
+///
+/// Assigning a value to a TrackingVH is always allowed, even if said TrackingVH
+/// no longer points to a valid value.
+template <typename ValueTy> class TrackingVH {
+  WeakVH InnerHandle;
 
-    // Check that this value is valid (i.e., it hasn't been deleted). We
-    // explicitly delay this check until access to avoid requiring clients to be
-    // unnecessarily careful w.r.t. destruction.
-    assert(ValueHandleBase::isValid(VP) && "Tracked Value was deleted!");
+public:
+  ValueTy *getValPtr() const {
+    // HLSL Change begin
+    // The original upstream change will assert here when accessing a TrackingVH
+    // is deleted.
+    //
+    // However, the llvm code that DXC forked has the implicit code like:
+    // TrackingVH V = nullptr;
+    //
+    // It will invoke setValPtr(nullptr) and then getValPtr(nullptr). So pull in
+    // the original upstream change in DXC will always assert here for debug
+    // build even this code is valid.
+    //
+    // The original upstream change works because of another upstream change
+    // https://github.com/llvm/llvm-project/commit/70a6051ddfd5f04777f2bc42503bb11bc8f1723a
+    // cleaned up the problematic code in DXC already.
```

Comment:
The other upstream change seems pretty minimal and safe.  If it could be merged to DXC without problems, wouldn't that be preferable to diverging in this change?